src/Lykke.AlgoStore.MatchingEngineAdapter.Services/Listening/ListeningService.cs

```diff
@@ -0,0 +1,162 @@
+
+        private readonly CancellationTokenSource _cts = new CancellationTokenSource();
+
+        private TcpListener _listener;
+
+        private bool _isDisposed;
+
+        /// <summary>
+        /// Initializes a <see cref="ListeningService"/>
+        /// </summary>
+        public ListeningService(
+            IAlgoClientInstanceRepository clientInstanceRepository,
+            IMessageHandler messageHandler,
+            [NotNull] ILog log,
+            ushort port)
+        {
+            _clientInstanceRepository = clientInstanceRepository ?? throw new ArgumentNullException(nameof(clientInstanceRepository));
+            _messageHandler = messageHandler ?? throw new ArgumentNullException(nameof(messageHandler));
+            _log = log ?? throw new ArgumentNullException(nameof(log));
+
+            _port = port;
+        }
+
+        /// <summary>
+        /// Starts listening and starts the connection accepting thread
+        /// </summary>
+        public async Task Start()
+        {
+            _listener = new TcpListener(IPAddress.Any, _port);
+            _listener.Server.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.ReuseAddress, true);
+            _listener.Start();
+
+            await AcceptConnections();
+        }
+
+        /// <summary>
+        /// Disposes this <see cref="ListeningService"/> by shutting down the accepting thread,
+        /// the listener and disposing the load balancers
+        /// </summary>
+        public void Stop()
+        {
+            Dispose();
+        }
+
+        /// <summary>
+        /// Disposes this <see cref="ListeningService"/> by shutting down the accepting thread,
+        /// the listener and disposing the load balancers
+        /// </summary>
+        public void Dispose()
+        {
+            Dispose(true);
+        }
+
+        protected virtual void Dispose(bool isDisposing)
+        {
+            if (_isDisposed) return;
+
+            if (isDisposing)
+            {
+                _listener.Stop();
+                _listener.Server.Close();
+                _listener.Server.Dispose();
+
+                var allConnections = _allWorkers.Keys.ToArray();
+
+                _cts.Cancel();
+
+                var allConnectionsClosedTask = Task.WhenAll(allConnections);
+                allConnectionsClosedTask.Wait();
+
+                _cts.Dispose();
+            }
+
+            _isDisposed = true;
+        }
+
+        /// <summary>
+        /// Accepts new connections until the listener is stopped
+        /// </summary>
+        private async Task AcceptConnections()
+        {
+            while (true)
+            {
+                var socket = await _listener.AcceptSocketAsync();
+
+                await _log.WriteInfoAsync(nameof(ListeningService), nameof(AcceptConnections),
+                    $"Accepting incoming connection from {socket.RemoteEndPoint}");
+
+                var networkStream = new NetworkStream(socket, true);
```

Comment:
Release resources

Reply:
The stream gets disposed inside the StreamWrapper